### PR TITLE
fixes to textarea-based nodes - resolve #441 #378

### LINKED ISSUE #441: Text input box sometimes collapses after editing text
collapsing text box

### PR DESCRIPTION
**Pull request by gmarinov**
screen shot 2016-02-15 at 12 03 02

### CODE REVIEW

File: browser/plugins/annotation.plugin.js [hunk 1]
```diff
@@ -8,36 +14,38 @@ var Annotation = E2.plugins.annotation = function() {
...
+		set text(str) 	{ return that.state.text = str},
+		get width()  	{ return that.state.width },
+		set width(px) 	{ return that.state.width = clamp(px, minWidth, maxWidth) },
+		get height()  	{ return that.state.height },
+		set height(px) 	{ return that.state.height = clamp(px, minHeight, maxHeight) }
+	}
+
+	var onChange 		= this.updateUi.bind(this)
+	var onBeginChange 	= this.beginBatchModifyState.bind(this)
+	var onEndChange 	= function() { that.endBatchModifyState('Annotation') }
```
Comment (kschzt) [inline browser/plugins/annotation.plugin.js]: DRY, maybe make like a `abstractTextInputPlugin` super class?

File: browser/plugins/const_text_generator.plugin.js [hunk 2]
```diff
@@ -20,34 +26,30 @@ Text.prototype.reset = function() {
...
+		get text()  	{ return that.state.text; },
+		set text(str) 	{ return that.state.text = str},
+		get width()  	{ return that.state.width },
+		set width(px) 	{ return that.state.width = clamp(px, minWidth, maxWidth) },
+		get height()  	{ return that.state.height },
+		set height(px) 	{ return that.state.height = clamp(px, minHeight, maxHeight) }
+	}
+
+	var onChange 		= this.updateUi.bind(this)
+	var onBeginChange 	= this.beginBatchModifyState.bind(this)
```
Comment (kschzt) [inline browser/plugins/const_text_generator.plugin.js]: DRY 2

File: browser/plugins/annotation.plugin.js (hunk 1 above)
Comment (gmarinov) [inline browser/plugins/annotation.plugin.js]: i was hoping to handle that in a separate PR. similar UI issues are building up.

File: browser/scripts/plugin.js [hunk 3]
```diff
@@ -48,5 +49,38 @@ Plugin.prototype.update_input = function(slot, data) {
 }
 
 
+Plugin.prototype.beginBatchModifyState = function() {
+	var that = this, state = that.state, oldState = that._oldState
+	if (oldState) {
+		Object.keys(oldState).forEach(function(prop) {
+			delete oldState[prop]
```
Comment (kschzt) [inline browser/scripts/plugin.js]: why not just reset `this._oldState` back to `{}`?

File: browser/scripts/plugin.js [hunk 4]
```diff
@@ -48,5 +49,38 @@ Plugin.prototype.update_input = function(slot, data) {
 }
 
 
+Plugin.prototype.beginBatchModifyState = function() {
+	var that = this, state = that.state, oldState = that._oldState
```
Comment (kschzt) [inline browser/scripts/plugin.js]: `that` is unnecessary

File: browser/scripts/plugin.js [hunk 5]
```diff
@@ -48,5 +49,38 @@ Plugin.prototype.update_input = function(slot, data) {
...
+	if (oldState) {
+		Object.keys(oldState).forEach(function(prop) {
+			delete oldState[prop]
+		})
+	} else {
+		oldState = this._oldState = {}
+	}
+
+	Object.keys(state).forEach(function(prop) {
```
Comment (kschzt) [inline browser/scripts/plugin.js]: consider `this._oldState = _.clone(this.state)`

File: browser/scripts/plugin.js [hunk 6]
```diff
@@ -48,5 +49,38 @@ Plugin.prototype.update_input = function(slot, data) {
...
+	if ( !oldState  ||  Object.keys(oldState).length === 0 ) return
+	E2.app.undoManager.begin(stepName)
+	Object.keys(state).forEach(function(prop) {
+		if (oldState[prop] !== state[prop] ) {
+			that.undoableSetState(prop, state[prop], oldState[prop])
+		}
+	})
+	E2.app.undoManager.end()
+
+	Object.keys(oldState).forEach(function(prop) {
```
Comment (kschzt) [inline browser/scripts/plugin.js]: could do a `this._oldState = {}`

File: browser/scripts/ui/uiTextArea.js [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+	})
+	container.addClass('uiTextArea')
+
+	textarea.css({
+		'font-size'	: '8.7pt',
+		'line-height': '1.45',
+		'border'	: borderPx + 'px solid #999',
+		'margin'	: '0 0 5px 0',
+		'padding'	: paddingPx + 'px',
+		'resize' 	: 'none',
```
Comment (kschzt) [inline browser/scripts/ui/uiTextArea.js]: shouldn't these be in the less?

File: browser/scripts/plugin.js (hunk 3 above)
Comment (gmarinov) [inline browser/scripts/plugin.js]: i'm preserving the object and any references it might have

**Comment by kschzt** [comment]
lgtm - please fix the DRY soon

**Pull request merged**